llvm/lib/Target/Mips/MipsAsmPrinter.cpp

```diff
@@ -329,6 +334,51 @@ static void emitDirectiveRelocJalr(const MachineInstr &MI,
   }
 }
 
+// NOTE: This is being used to control optimization of BALC in the linker.
+// TODO: Introduce a new relocation that will serve for this purpose only.
+static bool emitDirectiveRelocBalcStubs(const MachineInstr &MI,
+                                        MCContext &OutContext,
+                                        TargetMachine &TM,
+                                        MCStreamer &OutStreamer) {
+  assert(EmitNanoMipsBalcStubsReloc);
+  if (MI.getMF() && MI.getMF()->getFunction().hasOptSize()) {
+    MCSymbol *OffsetLabel = OutContext.createTempSymbol();
+    const MCExpr *OffsetExpr = MCSymbolRefExpr::create(OffsetLabel, OutContext);
+    // This symbol will be used to indicate that we do want balc stubs
```

Comment:
Strictly speaking we don't need a symbol for this relocation, nullptr will do. If you want to add one, add it once and reuse for the whole module.